### PR TITLE
FEAT: Improve pandas udf performance for many arguments

### PR DESCRIPTION
**Pull request by icexelloss**
Problem
--------

The existing pandas udf implementation suffers from huge performance problem when the udf takes many arguments. 

The problem is that, the existing implementation tries to register execution rules for every possible combination of argument types, resulting in an exponential growth execution rules (exponential to the number of udf arguments)

For an example, if a udf takes n arguments, it will ended register O(4^n) rules because each argument can take >4 different types, e.g., for floating numbers, (float, np.floating, pd.Series, pd.SeriesGroupBy) and the existing implementation tries to register the Cartesian product of those.

Proposed Solution
------------------
In this PR, the number of rules registered for a given number of arguments is reduced to 3. I.e., for a 10 argument UDF, only 3 rules will be registered. 
* (pd.SeriesGroupBy, pd.SeriesGroupBy, ..., pd.SeriesGroupBy)
* (pd.Series, pd.Series, ..., pd.Series) 
* (object, object, ..., object)

The last rule (object, object, ..., object) is a "catch all" rule to handle scalar inputs.

Additional Changes
--------------------
In this PR, I also take the chance to simplify pandas/core.py. This is because I found a bug in the existing implementation where pre_execute is not called for all nodes. 

Microbenchmark
-----------------
```
import numpy as np
import pandas as pd
import ibis
import ibis.expr.datatypes as dt
from ibis.pandas.udf import udf
import time

df = pd.DataFrame(
    {
        'a': np.arange(4, dtype=float).tolist()
        + np.random.rand(3).tolist(),
        'b': np.arange(4, dtype=float).tolist()
        + np.random.rand(3).tolist(),
        'c': np.arange(7, dtype=int).tolist(),
        'key': list('ddeefff'),
    }
)

client = ibis.pandas.connect({'table': df})

t2 = client.table('table')

@udf.elementwise(input_type=[dt.double] * 6, output_type=dt.double)
def my_udf(
    c1,
    c2,
    c3,
    c4,
    c5,
    c6
):
    return c1

expr = my_udf(*([t2.a] * 6))

start = time.time()
result = expr.execute()

print("Elapsed time:", time.time() - start)
```

Before the PR:
```
Elapsed time: 0.1456921100616455
```

After the PR:
```
Elapsed time: 0.0021181106567382812
```

### CODE REVIEW

**Comment by xmnlab** [comment]
@icexelloss will this PR also fix the udf issues related the 2 tests that were failing?  ref https://github.com/ibis-project/ibis/issues/2065

**Comment by icexelloss** [comment]
I don't believe so. I didn't address any window related issues. I will take a look at that separately .

**Comment by xmnlab** [comment]
sounds good! thanks @icexelloss !

File: ibis/pandas/udf.py [hunk 1]
```diff
@@ -542,16 +542,9 @@ def pre_execute_elementwise_udf(
...
-    ]
-
-    @toolz.compose(
-        *(
-            execute_node.register(ops.ElementWiseVectorizedUDF, *types)
-            for types in group_by_signatures
-        )
+
+    @execute_node.register(
+        ops.ElementWiseVectorizedUDF, *(itertools.repeat(SeriesGroupBy, nargs))
```
Comment (emilyreff7) [inline ibis/pandas/udf.py]: Why do we not also use the object rule here for a 'scalar input catch-all'?

Comment (icexelloss) [reply ibis/pandas/udf.py]: Because we only need one catch all for `ops.ElementWiseVectorizedUDF`

We do not need another here because both of these register rule for `ops.ElementWiseVectorizedUDF`

File: ibis/pandas/core.py [hunk 2]
```diff
@@ -214,55 +214,6 @@ def execute_until_in_scope(
...
-
-    Returns
-    -------
-    result : Mapping[
-        ibis.expr.operations.Node,
-        Union[pandas.Series, pandas.DataFrame, scalar_types]
-    ]
-        A mapping from node to the computed result of that Node
-    """
-    assert post_execute_ is not None, 'post_execute_ is None'
```
[the remark this answers is not in the document]
Comment (icexelloss) [reply ibis/pandas/core.py]: Good catch. This is accidental. Let me add it back in `execute_until_in_scope`

File: ibis/pandas/udf.py [hunk 3]
```diff
@@ -529,10 +528,8 @@ def wrapped(*args):
 
 
 @pre_execute.register(ops.ElementWiseVectorizedUDF)
-@pre_execute.register(ops.ElementWiseVectorizedUDF, PandasClient)
-def pre_execute_elementwise_udf(
-    op, *clients, scope=None, aggcontet=None, **kwargs
-):
+@pre_execute.register(ops.ElementWiseVectorizedUDF, ibis.client.Client)
```
Comment (jreback) [inline ibis/pandas/udf.py]: doesn't this catch more generically here? I guess we don't have ElementWiseVector UDFs on other clients atm, but why this change?

Comment (icexelloss) [reply ibis/pandas/udf.py]: This isn't really necessary but follows the same convention as https://github.com/ibis-project/ibis/blob/master/ibis/pandas/dispatch.py#L49

File: ibis/pandas/core.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (icexelloss) [reply ibis/pandas/core.py]: The check is in `execute_until_in_scope` https://github.com/ibis-project/ibis/blob/ee8ca9f1c002a153cce1e9cc8caa90374420bc52/ibis/pandas/core.py#L212

**Comment by jreback** [comment]
thanks @icexelloss 

did we have an associated issue to close?
